<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#工具包导入&amp;数据读取" data-toc-modified-id="工具包导入&amp;数据读取-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>工具包导入&amp;数据读取</a></span><ul class="toc-item"><li><span><a href="#工具包导入" data-toc-modified-id="工具包导入-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>工具包导入</a></span></li><li><span><a href="#数据读取" data-toc-modified-id="数据读取-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>数据读取</a></span></li></ul></li><li><span><a href="#模型Baseline" data-toc-modified-id="模型Baseline-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>模型Baseline</a></span><ul class="toc-item"><li><span><a href="#模型训练框架" data-toc-modified-id="模型训练框架-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>模型训练框架</a></span></li><li><span><a href="#模型训练&amp;提交" data-toc-modified-id="模型训练&amp;提交-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>模型训练&amp;提交</a></span><ul class="toc-item"><li><span><a href="#原始数据" data-toc-modified-id="原始数据-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>原始数据</a></span><ul class="toc-item"><li><span><a href="#模型训练（CV:0.897267399）" data-toc-modified-id="模型训练（CV:0.897267399）-2.2.1.1"><span class="toc-item-num">2.2.1.1&nbsp;&nbsp;</span>模型训练（CV:0.897267399）</a></span></li><li><span><a href="#模型测试（PB：0.89558，-LB：0.89825,-5464/9038）" data-toc-modified-id="模型测试（PB：0.89558，-LB：0.89825,-5464/9038）-2.2.1.2"><span class="toc-item-num">2.2.1.2&nbsp;&nbsp;</span>模型测试（<font color="red">PB：0.89558， LB：0.89825, 5464/9038</font>）</a></span></li></ul></li><li><span><a href="#加入frequency-编码" data-toc-modified-id="加入frequency-编码-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>加入frequency 编码</a></span><ul class="toc-item"><li><span><a href="#模型训练(CV:0.909231)" data-toc-modified-id="模型训练(CV:0.909231)-2.2.2.1"><span class="toc-item-num">2.2.2.1&nbsp;&nbsp;</span>模型训练(CV:0.909231)</a></span></li><li><span><a href="#模型测试（PB：0.89778-，-LB：0.90043,-4563/9038）" data-toc-modified-id="模型测试（PB：0.89778-，-LB：0.90043,-4563/9038）-2.2.2.2"><span class="toc-item-num">2.2.2.2&nbsp;&nbsp;</span>模型测试（<font color="red">PB：0.89778 ， LB：0.90043, 4563/9038</font>）</a></span></li></ul></li><li><span><a href="#换个参数换个运气" data-toc-modified-id="换个参数换个运气-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>换个参数换个运气</a></span><ul class="toc-item"><li><span><a href="#模型训练(CV:0.90244)" data-toc-modified-id="模型训练(CV:0.90244)-2.2.3.1"><span class="toc-item-num">2.2.3.1&nbsp;&nbsp;</span>模型训练(CV:0.90244)</a></span></li><li><span><a href="#模型测试（PB：0.89914--，-LB：0.90037,-3271/9038）" data-toc-modified-id="模型测试（PB：0.89914--，-LB：0.90037,-3271/9038）-2.2.3.2"><span class="toc-item-num">2.2.3.2&nbsp;&nbsp;</span>模型测试（<font color="red">PB：0.89914  ， LB：0.90037, 3271/9038</font>）</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#小结" data-toc-modified-id="小结-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>小结</a></span></li></ul></div>

# 工具包导入&数据读取
## 工具包导入

In [2]:
import numpy as np # linear algebra
import pandas as pd
import seaborn as sns
import time
import matplotlib.pyplot as plt
from sklearn.model_selection import StratifiedKFold
import lightgbm as lgb
from sklearn import metrics
from scipy.optimize import curve_fit
from scipy import asarray as ar,exp
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import roc_auc_score, roc_curve
from tqdm import tqdm_notebook as tqdm

## 数据读取 

In [3]:
path = '/data/Data_JieZhang/santander-customer-transaction-prediction/'
train = pd.read_csv(path + 'train.csv')
test  = pd.read_csv(path + 'test.csv')

# 模型Baseline
## 模型训练框架

In [4]:
import lightgbm as lgb
from sklearn.model_selection import KFold,StratifiedKFold
def validation_prediction_lgb(X,y,feature_names, ratio =1, X_test = None,istest = False):
    n_fold = 5
    folds = StratifiedKFold(n_splits=n_fold, shuffle=True, random_state=42)
    params = {
        'num_leaves': 10,
        'max_bin': 127,
        'min_data_in_leaf': 11,
        'learning_rate': 0.02,
        'min_sum_hessian_in_leaf': 0.00245,
        'bagging_fraction': 1, 
        'bagging_freq': 5, 
        'feature_fraction': 0.5,
        'lambda_l1': 4.972,
        'lambda_l2': 2.276,
        'min_gain_to_split': 0.65,
        'max_depth': 32,
        'save_binary': False,
        'seed': 1337,
        'feature_fraction_seed': 1337,
        'bagging_seed': 1337,
        'drop_seed': 1337,
        'data_random_seed': 1337,
        'objective': 'binary',
        'boosting_type': 'gbdt',
        'verbose': 0,
        'metric': 'auc',
        'is_unbalance': False,
        'boost_from_average': False,
    } 

    importances = pd.DataFrame() 
    
    if istest:
        prediction = np.zeros(len(X_test))
    models = []
    
    for fold_n, (train_index, valid_index) in enumerate(folds.split(X,y)):
        print('Fold', fold_n, 'started at', time.ctime())
        X_train, X_valid = X[train_index], X[valid_index]
        y_train, y_valid = y[train_index], y[valid_index]
        weights = [ratio  if val == 1 else 1 for val in y_train]
        
        train_data = lgb.Dataset(X_train, label=y_train,  weight=weights)
        valid_data = lgb.Dataset(X_valid, label=y_valid)

        model = lgb.train(params,train_data,num_boost_round=20000,
                        valid_sets = [train_data, valid_data],verbose_eval=200,early_stopping_rounds = 200)
        
        imp_df = pd.DataFrame() 
        imp_df['feature']  = feature_names
        imp_df['split']    = model.feature_importance()
        imp_df['gain']     = model.feature_importance(importance_type='gain')
        imp_df['fold']     = fold_n + 1
        
        importances = pd.concat([importances, imp_df], axis=0)
        
        models.append(model)
        if istest == True:
            prediction += model.predict(X_test, num_iteration=model.best_iteration)/5
    if istest == True:     
        return models,importances, prediction
    else:
        return models,importances

## 模型训练&提交
### 原始数据
#### 模型训练（CV:0.897267399）

In [5]:
train_cols = [col for col in train.columns if col not in ['target','ID_code']]
print(len(train_cols))
models,importances,prediction = validation_prediction_lgb(train[train_cols].values, train['target'].values, train_cols, X_test= test[train_cols].values, istest=True)

200
Fold 0 started at Fri Apr 12 14:00:07 2019
Training until validation scores don't improve for 200 rounds.
[200]	training's auc: 0.817275	valid_1's auc: 0.801333
[400]	training's auc: 0.863716	valid_1's auc: 0.842882
[600]	training's auc: 0.885386	valid_1's auc: 0.860301
[800]	training's auc: 0.898083	valid_1's auc: 0.870158
[1000]	training's auc: 0.906955	valid_1's auc: 0.876647
[1200]	training's auc: 0.913327	valid_1's auc: 0.881143
[1400]	training's auc: 0.918267	valid_1's auc: 0.8846
[1600]	training's auc: 0.922355	valid_1's auc: 0.887141
[1800]	training's auc: 0.925715	valid_1's auc: 0.88911
[2000]	training's auc: 0.928656	valid_1's auc: 0.89067
[2200]	training's auc: 0.931047	valid_1's auc: 0.892012
[2400]	training's auc: 0.933181	valid_1's auc: 0.893084
[2600]	training's auc: 0.935053	valid_1's auc: 0.893985
[2800]	training's auc: 0.936765	valid_1's auc: 0.89471
[3000]	training's auc: 0.938306	valid_1's auc: 0.895195
[3200]	training's auc: 0.939759	valid_1's auc: 0.895654
[34

In [6]:
np.mean([0.896561, 0.89753, 0.90264, 0.89641, 0.893196])

0.8972673999999999

#### 模型测试（<font color=red>PB：0.89558， LB：0.89825, 5464/9038</font>）
- PB：0.89558,LB：0.89825, 线上排名:5464/9038

![](./pic/baseline_score.png)

通过线下CV的结果和线上提交的结果来看,线上线下还是一致的。


In [8]:
sub            = pd.DataFrame()
sub['ID_code'] = test['ID_code'].values
sub['target']  = prediction
sub.to_csv('baseline.csv', index = None)

### 加入frequency 编码

- frequency编码可以认为是一种流行度的编码,经常是用在类别型的变量上,本次比赛大量的数据看上去是数值型的, 不过貌似现在都已经习惯只要是匿名的比赛，不管三七二十一先直接上去试一下。


In [9]:
tr_te = pd.concat([train, test], axis=0, ignore_index=True)

In [10]:
for col in tqdm(train.columns):
    if 'var' in col:
        tr_te[col + '_cnt'] = tr_te[col].map(tr_te[col].value_counts())

HBox(children=(IntProgress(value=0, max=202), HTML(value='')))

#### 模型训练(CV:0.909231)

In [16]:
train_cols = [col for col in tr_te.columns if col not in ['target','ID_code']]
print(len(train_cols))
models2,importances2,prediction2 = validation_prediction_lgb(tr_te.iloc[:train.shape[0],:][train_cols].values, tr_te.iloc[:train.shape[0],:]['target'].values, train_cols, X_test= tr_te.iloc[train.shape[0]:,:][train_cols].values, istest=True)

400
Fold 0 started at Fri Apr 12 15:02:53 2019
Training until validation scores don't improve for 200 rounds.
[200]	training's auc: 0.819658	valid_1's auc: 0.804276
[400]	training's auc: 0.867814	valid_1's auc: 0.847607
[600]	training's auc: 0.889196	valid_1's auc: 0.86566
[800]	training's auc: 0.902454	valid_1's auc: 0.876633
[1000]	training's auc: 0.911545	valid_1's auc: 0.883578
[1200]	training's auc: 0.918043	valid_1's auc: 0.888176
[1400]	training's auc: 0.923219	valid_1's auc: 0.891737
[1600]	training's auc: 0.927386	valid_1's auc: 0.894741
[1800]	training's auc: 0.930817	valid_1's auc: 0.897007
[2000]	training's auc: 0.933882	valid_1's auc: 0.899048
[2200]	training's auc: 0.936366	valid_1's auc: 0.900616
[2400]	training's auc: 0.938614	valid_1's auc: 0.901945
[2600]	training's auc: 0.940627	valid_1's auc: 0.903205
[2800]	training's auc: 0.942426	valid_1's auc: 0.904284
[3000]	training's auc: 0.94413	valid_1's auc: 0.905132
[3200]	training's auc: 0.945695	valid_1's auc: 0.90582
[

[5400]	training's auc: 0.960411	valid_1's auc: 0.907476
[5600]	training's auc: 0.961549	valid_1's auc: 0.90763
[5800]	training's auc: 0.962686	valid_1's auc: 0.907656
[6000]	training's auc: 0.963773	valid_1's auc: 0.907754
[6200]	training's auc: 0.964841	valid_1's auc: 0.907779
[6400]	training's auc: 0.965855	valid_1's auc: 0.907896
[6600]	training's auc: 0.966881	valid_1's auc: 0.907934
[6800]	training's auc: 0.967847	valid_1's auc: 0.908053
[7000]	training's auc: 0.9688	valid_1's auc: 0.908124
Early stopping, best iteration is:
[6930]	training's auc: 0.968478	valid_1's auc: 0.908156
Fold 4 started at Fri Apr 12 15:12:20 2019
Training until validation scores don't improve for 200 rounds.
[200]	training's auc: 0.823471	valid_1's auc: 0.804201
[400]	training's auc: 0.870609	valid_1's auc: 0.84386
[600]	training's auc: 0.891305	valid_1's auc: 0.860703
[800]	training's auc: 0.903767	valid_1's auc: 0.870909
[1000]	training's auc: 0.912172	valid_1's auc: 0.877787
[1200]	training's auc: 0.91

In [18]:
np.mean([0.909266,0.908574,0.915016,0.908156,0.905148])

0.9092319999999999

In [17]:
sub            = pd.DataFrame()
sub['ID_code'] = test['ID_code'].values
sub['target']  = prediction2
sub.to_csv('baseline_frequency.csv', index = None)

#### 模型测试（<font color=red>PB：0.89778 ， LB：0.90043, 4563/9038</font>）
- PB：0.89778 ， LB：0.90043, 线上排名:4563/9038

![](./pic/baseline_frequency_score.png)

通过线下CV的结果和线上提交的结果来看,线上线下出现了不一致的情况,这究竟是什么情况呢?


### 换个参数换个运气

In [23]:
import lightgbm as lgb
from sklearn.model_selection import KFold,StratifiedKFold
def validation_prediction_lgb2(X,y,feature_names, ratio =1, X_test = None,istest = False):
    n_fold = 5
    folds = StratifiedKFold(n_splits=n_fold, shuffle=True, random_state=42)
      
    params = {
    'bagging_freq': 5,
    'bagging_fraction': 0.335,
    'boost_from_average':'false',
    'boost': 'gbdt',
    'feature_fraction': 0.041,
    'learning_rate': 0.0083,
    'max_depth': -1,
    'metric':'auc',
    'min_data_in_leaf': 80,
    'min_sum_hessian_in_leaf': 10.0,
    'num_leaves': 13,
    'num_threads': 8,
    'tree_learner': 'serial',
    'objective': 'binary',
    'verbosity': 1}

    importances = pd.DataFrame() 
    
    if istest:
        prediction = np.zeros(len(X_test))
    models = []
    
    for fold_n, (train_index, valid_index) in enumerate(folds.split(X,y)):
        print('Fold', fold_n, 'started at', time.ctime())
        X_train, X_valid = X[train_index], X[valid_index]
        y_train, y_valid = y[train_index], y[valid_index]
        weights = [ratio  if val == 1 else 1 for val in y_train]
        
        train_data = lgb.Dataset(X_train, label=y_train,  weight=weights)
        valid_data = lgb.Dataset(X_valid, label=y_valid)

        model = lgb.train(params,train_data,num_boost_round=20000,
                        valid_sets = [train_data, valid_data],verbose_eval=200,early_stopping_rounds = 200)
        
        imp_df = pd.DataFrame() 
        imp_df['feature']  = feature_names
        imp_df['split']    = model.feature_importance()
        imp_df['gain']     = model.feature_importance(importance_type='gain')
        imp_df['fold']     = fold_n + 1
        
        importances = pd.concat([importances, imp_df], axis=0)
        
        models.append(model)
        if istest == True:
            prediction += model.predict(X_test, num_iteration=model.best_iteration)/5
    if istest == True:     
        return models,importances, prediction
    else:
        return models,importances

#### 模型训练(CV:0.90244)

In [24]:
train_cols = [col for col in tr_te.columns if col not in ['target','ID_code']]
print(len(train_cols))
models3,importances3,prediction3 = validation_prediction_lgb2(tr_te.iloc[:train.shape[0],:][train_cols].values, tr_te.iloc[:train.shape[0],:]['target'].values, train_cols, X_test= tr_te.iloc[train.shape[0]:,:][train_cols].values, istest=True)

400
Fold 0 started at Fri Apr 12 16:19:46 2019
Training until validation scores don't improve for 200 rounds.
[200]	training's auc: 0.877964	valid_1's auc: 0.860445
[400]	training's auc: 0.89025	valid_1's auc: 0.87212
[600]	training's auc: 0.895412	valid_1's auc: 0.876649
[800]	training's auc: 0.899266	valid_1's auc: 0.880342
[1000]	training's auc: 0.901961	valid_1's auc: 0.88194
[1200]	training's auc: 0.904812	valid_1's auc: 0.884495
[1400]	training's auc: 0.906813	valid_1's auc: 0.886337
[1600]	training's auc: 0.908931	valid_1's auc: 0.888102
[1800]	training's auc: 0.910851	valid_1's auc: 0.889523
[2000]	training's auc: 0.912521	valid_1's auc: 0.890787
[2200]	training's auc: 0.91415	valid_1's auc: 0.891816
[2400]	training's auc: 0.915727	valid_1's auc: 0.892783
[2600]	training's auc: 0.917129	valid_1's auc: 0.8937
[2800]	training's auc: 0.918491	valid_1's auc: 0.894353
[3000]	training's auc: 0.919811	valid_1's auc: 0.894987
[3200]	training's auc: 0.921146	valid_1's auc: 0.89566
[3400

[9000]	training's auc: 0.944911	valid_1's auc: 0.90769
[9200]	training's auc: 0.945566	valid_1's auc: 0.907757
[9400]	training's auc: 0.94622	valid_1's auc: 0.907761
[9600]	training's auc: 0.946899	valid_1's auc: 0.907865
[9800]	training's auc: 0.947587	valid_1's auc: 0.907909
[10000]	training's auc: 0.948256	valid_1's auc: 0.908008
[10200]	training's auc: 0.948935	valid_1's auc: 0.908079
[10400]	training's auc: 0.949545	valid_1's auc: 0.908161
[10600]	training's auc: 0.950209	valid_1's auc: 0.908185
[10800]	training's auc: 0.950823	valid_1's auc: 0.908239
Early stopping, best iteration is:
[10767]	training's auc: 0.950724	valid_1's auc: 0.908253
Fold 3 started at Fri Apr 12 16:30:26 2019
Training until validation scores don't improve for 200 rounds.
[200]	training's auc: 0.877577	valid_1's auc: 0.857795
[400]	training's auc: 0.888396	valid_1's auc: 0.868991
[600]	training's auc: 0.894312	valid_1's auc: 0.873986
[800]	training's auc: 0.899061	valid_1's auc: 0.878024
[1000]	training's a

In [25]:
np.mean([0.901866,0.902338,0.908253, 0.901488,0.898279 ])

0.9024447999999999

In [26]:
sub            = pd.DataFrame()
sub['ID_code'] = test['ID_code'].values
sub['target']  = prediction3
sub.to_csv('baseline_frequency2.csv', index = None)

#### 模型测试（<font color=red>PB：0.89914  ， LB：0.90037, 3271/9038</font>）
- PB：0.89914 , LB：0.90037, 线上排名:3271/9038

![](./pic/baseline_frequency_score2.png)

果然换了参数, 线上LB虽然没变,但是PB上升了1000多名。

# 小结

在这一章，我们：
- 为了快速验证线下与线上的差距,直接采用原始数据集的特征训练了5折的LGB,并提交,发现线上线下基本是一致的;
- 后来,我们直接无脑的对所有数据进行频率编码, 可惜问题出现了,我们发现线下和线上出现了不一致的情况,不过好处是线下和线上都提升了;坏处是线下提升的幅度和线上是不一致的; 虽然我个人是不推荐实践中这么做,但是如果做匿名的比赛，这倒也是一个非常好的尝试;
- 最后,我们随机替换了一组LGB的参数,PB成绩就上升了1000名,这也不奇怪,开源的Kernel里面已经说了,通过调整参数原始数据LB都可以达到0.9,这个比我们上面的0.898要高了很多(在这个比赛中....);

本文的结果在比赛开赛前一个月还可以在前100名,但是在后期大家发现magic之后,这分数就是铜牌开外了,究竟是什么样神奇的magic呢,在下一篇文章中,我会给出本次比赛最重要的magic之一,并做简单的解释,欢迎大家继续关注我的文章。
